src/main/java/org/opensearch/neuralsearch/mappingtransformer/SemanticMappingTransformer.java

```diff
@@ -0,0 +1,429 @@
+                        listener.onFailure(new RuntimeException(String.join("; ", errors)));
+                    }
+                }
+            }, e -> {
+                hasError.set(true);
+                errors.add(e.getMessage());
+                if (counter.decrementAndGet() == 0) {
+                    listener.onFailure(new RuntimeException(String.join("; ", errors)));
+                }
+            }));
+        }
+    }
+
+    private void modifyMappings(
+        @NonNull final Map<String, MLModel> modelIdToConfigMap,
+        @NonNull final Map<String, Object> mappings,
+        @NonNull final Map<String, List<String>> modelIdToFieldPathMap,
+        @NonNull final Map<String, Map<String, Object>> semanticFieldPathToConfigMap
+    ) {
+        for (String modelId : modelIdToFieldPathMap.keySet()) {
+            final MLModel mlModel = modelIdToConfigMap.get(modelId);
+            final List<String> fieldPathList = modelIdToFieldPathMap.get(modelId);
+            for (String fieldPath : fieldPathList) {
+                try {
+                    final Map<String, Object> semanticInfoConfig = createSemanticInfoField(mlModel);
+                    final Map<String, Object> fieldConfig = semanticFieldPathToConfigMap.get(fieldPath);
+                    setSemanticInfoField(mappings, fieldPath, fieldConfig.get(SEMANTIC_INFO_FIELD_NAME), semanticInfoConfig);
+                } catch (IllegalArgumentException e) {
+                    throw new IllegalArgumentException(getModifyMappingErrorMessage(fieldPath, e.getMessage()), e);
+                } catch (Exception e) {
+                    throw new RuntimeException(getModifyMappingErrorMessage(fieldPath, e.getMessage()), e);
+                }
+            }
+        }
+    }
+
+    private String getModifyMappingErrorMessage(@NonNull final String fieldPath, final String cause) {
+        return String.format(Locale.ROOT, "Failed to transform the mapping for the semantic field at %s due to %s", fieldPath, cause);
+    }
+
+    @VisibleForTesting
+    private Map<String, Object> createSemanticInfoField(final @NonNull MLModel modelConfig) {
+        final Map<String, Object> embeddingConfig = switch (modelConfig.getAlgorithm()) {
+            case FunctionName.TEXT_EMBEDDING -> createEmbeddingConfigForTextEmbeddingModel(modelConfig);
+            case FunctionName.SPARSE_ENCODING, FunctionName.SPARSE_TOKENIZE -> getBaseSparseEmbeddingConfig();
+            case FunctionName.REMOTE -> createEmbeddingConfigForRemoteModel(modelConfig);
+            default -> throw new IllegalArgumentException(
+                String.format(
+                    Locale.ROOT,
+                    "The algorithm %s of the model %s is not supported in the semantic field. Supported algorithms: [%s].",
+                    modelConfig.getAlgorithm().name(),
+                    modelConfig.getModelId(),
+                    String.join(",", SUPPORTED_MODEL_ALGORITHMS)
```

Comment:
nit: you may not need to use `String.join` since Set toString() already print elements with comma delimited.